### ISSUE TITLE
New Binary/String type

### ISSUE THREAD

**Issue by ritchie46** [pola-rs/polars#13459]
The goal is to replace the current Arrow (Large)String type with a string type that allows a union between an inlined small string and an offset to a string that is allocated somewhere else.

This would prevent the terrible performance we have when filtering/gathering large string data as that forces a copy of all bytes. Second this type also allows string interning. As duplicates can only be stored once in the buffer and then we can point to that string multiple times.

Relevant arrow discussion here: https://lists.apache.org/thread/w88tpz76ox8h3rxkjl4so6rg3f1rv7wt

- implement in `polars-arrow` #13243 
- use prefix in equality
- implement IPC/Parquet
- migrate polars and all compute

**Comment by Steiniche** [comment]
Hi @ritchie46 , I believe that this issue can be closed and the functionality is merged and released.

**Comment by adriangb** [comment]
Is this available in the Python API? I don't see any references to the type.

**Comment by ritchie46** [comment]
It is available. Our string column type is backed by this. Polars keeps it simple, we have 1 string type, and you use it automatically.

**Comment by adriangb** [comment]
If I understand correctly this type can/should intern strings. On Polars `1.4.1` I tried comparing to a Categorical column and get a _much_ smaller size for the categorical column:

```python
import polars as pl

large_string = 'a' * 10_000
data = [large_string] * 100_000

df = pl.DataFrame({'x': data}, schema={'x': pl.String})
print(df.estimated_size())  # 1000000000

df = pl.DataFrame({'x': data}, schema={'x': pl.Categorical})
print(df.estimated_size())  # 410000
```

Is this expected? Am I misunderstanding the type?

**Comment by ritchie46** [comment]
Yes we can. Though there is cost in interning checking upon construction. Currently we only do that if you extend internally. `series.new_from_index`, I believe.

Later we will also look into interning checking until a limited size is reached.

**Comment by adriangb** [comment]
Indeed:

```python
import polars as pl

large_string = 'a' * 10_000
data = [large_string] * 100_000

df = pl.DataFrame({'x': data}, schema={'x': pl.String})
assert df.shape[0] == 100_000
print(df.estimated_size())  # 1000000000

df = pl.DataFrame({'x': data}, schema={'x': pl.Categorical})
assert df.shape[0] == 100_000
print(df.estimated_size())  # 410000

s = pl.Series('x', [large_string], pl.String)
s = s.new_from_index(0, len(data))
df = pl.DataFrame({'x': s})
assert df.shape[0] == 100_000
print(df.estimated_size())  # 10000
```

It's even better than a dictionary / categorical column!

Agreed it would be nice to heuristically do this when constructing, or at least to be able to force it on if I know my data has a lot of duplication of large strings.
